Add Code Insights Dashboard page UI

vovakulikov opened the pull request:
Closes https://github.com/sourcegraph/sourcegraph/issues/22225

### Background

**Adding dashboard UI** implies implementation of select/listbox UI, without dashboard creation UI. Users should be able to switch between different dashboards (All, personal, org level dashboards) via select dashboard UI. [You can find designs for this UI here](https://www.figma.com/file/g5lSlPBnHt00wCogjRFnt5/RFC-398%3A-Easily-Viewing-and-Sharing-Code-Insights-WIP?node-id=275%3A1628)

![Group 356](https://user-images.githubusercontent.com/18492575/123830137-b2a5a580-d90b-11eb-9cf9-b696747e8d40.png)


What have been done 
- [x] Change built-in dashboards paradigm on dashboards levels 
Example:
        - **All**
        - **Personal**
              - Built-in personal dashboard with all insights
              - Custom personal dashboards
        - **Organization 1**
           - Built-in Organization 1 dashboard with all insights withing this org
           - Custom Organization 1 dashboards
        - **Organization N**
           ...
- [x] Imlement listbox dashboard component

The following two points will be resolved in the next PR, so as not to increase the size of this PR https://github.com/sourcegraph/sourcegraph/pull/22474
- [ ] Connect UI logic and dashboard data 
- [ ] Connect UI logic to URL routes and sync active dashboard with insights route URL

vovakulikov commented:
**UI aspects** 

I implemented almost all details of the dashboard select component according to our designs, however, there are still some small questions to address 
- Visual state for the selected option (now I left this as bolded text just because designs selected state looks like disabled option) 
- Color for hover state of options (currently it is just blue color from reach UI styles, I saw that we use this for other context menu components but just double check Do we need to use our own color here?)
- Also I put badges to the right in select options for better readability but don't have a strong opinion about this.

<img width="401" alt="Screenshot 2021-06-30 at 14 01 26" src="https://user-images.githubusercontent.com/18492575/123950040-efc17480-d9ab-11eb-9e9b-2c93672cbd68.png">

@AlicjaSuska I added a storybook story for better UI review take a look when you have a minute

AlicjaSuska commented:
@vovakulikov thank you for sharing this! Looks great already. Here are some notes/answers to your questions:

- Thank you for catching the selected inconsistency with other menus. I made the corrections on this [mockup](https://www.figma.com/file/g5lSlPBnHt00wCogjRFnt5/RFC-398-Easily-Viewing-and-Sharing-Code-Insights-WIP?node-id=846%3A68). Hover/focus should be blue. The selected dashboard should use background-3. Please, do not use bold text.
- I've noticed that the dropdown is a little narrower in the storybook than in the designs. I want to make sure that is will be as wide as the Insight box (so it aligns with our columns layout)
- Thank you for adding tooltips to the long dashboard names that are not visible. Would it be possible to add tooltips also to badges that are too long to fit?
- I need to think over the placement of the badges. I see why you've added them on the right. Let me get back to you :)

AlicjaSuska reviewed (changes requested):
Please, see the comment above. Thank you

vovakulikov commented:
@AlicjaSuska thanks for the UI review. 

1. I've changed the selected visual state for the select option component. I used
 `background-color: var(--color-bg-2);` because `color-bg-3` is used in the secondary badge component (I mean maybe this color has a different name but the color value is the same)
 
<img width="391" alt="Screenshot 2021-07-01 at 10 53 21" src="https://user-images.githubusercontent.com/18492575/124087491-b85dd100-da5a-11eb-9b66-84ebee1aa9c6.png">

Should we use a different bootstrap color for the badge here? 

2. I fixed the width value for the select component. Designs layout has 364px width for this component = 364/16 = 22.75rem. Thanks for catching this.

<img width="457" alt="Screenshot 2021-07-01 at 10 59 42" src="https://user-images.githubusercontent.com/18492575/124088265-75502d80-da5b-11eb-828e-a58863af6f97.png">

3. Now badge also has a native title tooltip.

valerybugakov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/components/menu-button/MenuButton.tsx:
```diff
@@ -0,0 +1,73 @@
+                {badge && <Badge value={badge} className={classnames('ml-1 mr-1', styles.listboxButtonBadge)} />}
+            </span>
+
+            {isExpanded ? (
+                <ChevronUpIcon className={styles.listboxButtonIcon} />
+            ) : (
+                <ChevronDownIcon className={styles.listboxButtonIcon} />
+            )}
```
nit: this is minor but important for more complex cases — it ensures that the same set of props is passed to the same UI element.

```tsx
const ListboxButtonIcon = isExpanded ? ChevronUpIcon : ChevronDownIcon

return (
  <ListboxButtonIcon className={styles.listboxButtonIcon} />
)

valerybugakov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/components/menu-button/MenuButton.tsx:
```diff
@@ -0,0 +1,73 @@
+                const dashboard = dashboards.find(dashboard => dashboard.id === value)
+
+                if (!dashboard) {
+                    return <MenuButtonContent title="Unknown value" isExpanded={isExpanded} />
```
Is it an error case that we should care about, maybe report through logger?

vovakulikov replied:
Yeah. This is a possible case because we can pass any dashboard id in URL and if we didn't find anything by this ID we get an Unknown value. 

I think logger is a good idea. I would prefer to add this in this PR https://github.com/sourcegraph/sourcegraph/pull/22474

and something on URL parsing step rather than in render method. Thanks for the suggestion.

valerybugakov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/components/select-option/SelectOption.module.scss:
```diff
@@ -0,0 +1,30 @@
+    }
+
+    &__badge {
+        max-width: 8rem;
```
Is this related to the badge max-width?

```scss
    &__badge {
        max-width: 8rem;
```

If so, the CSS variable would be good here to ensure that these values are in sync.

vovakulikov replied:
Do you mean the select option badge and the menu button badge? I was thinking about this too. But the problem here is that I can't have CSS variable here on `DashboardSelect.module` level just because this select panel is rendered in the portal, So that means that we would have to put this variable on the root level which is not good I believe because it seems like this variable is useful only for the dashboard select component only.

Like 
```css
:root {
 --code-insight-dashboard-select-badge-max-width: 8rem
}
```

valerybugakov replied:
Oh, I see. Maybe then a small comment would be good to avoid forgetting to update this value in both places.

valerybugakov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/components/badge/Badge.tsx:
```diff
@@ -0,0 +1,19 @@
+    className?: string
+}
+
+export const Badge: React.FunctionComponent<BadgeProps> = props => {
```
We definitely need to update our shared `Badge` component so that it can be used in places like this one.

vovakulikov replied:
Yeah. I see that our existing badge covers only case about statuses. Badge here probably is bad naming from my side. I think `Tag` is a more appropriate name for this component here.

valerybugakov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/DashboardSelect.tsx:
```diff
@@ -0,0 +1,112 @@
+import { SelectDashboardOption, SelectOption } from './components/select-option/SelectOption'
+import styles from './DashboardSelect.module.scss'
+
+const LABEL_ID = 'insights-dashboards--select'
```
Curious, why is this id extracted into a separate variable and required on the `VisuallyHidden` element?

vovakulikov replied:
My bad, I forgot to add a visually hidden element here for voice-over support. I did it in the next PR about dashboard select here https://github.com/sourcegraph/sourcegraph/pull/22474/files#diff-76e93e914a8a7561770131d3053091e15d954200943afc00c3db396531039d0fR45

valerybugakov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/DashboardSelect.tsx:
```diff
@@ -0,0 +1,112 @@
+        setValue(value)
+    }
+
+    const organizationGroups = useMemo(() => getDashboardOrganizationsGroups(dashboards), [dashboards])
```
Why do we need the optimization with `useMemo` here?

vovakulikov replied:
My bad habit, I think this isn't the case for a big number of organizations here. But I unknowingly added memoization here to avoid regrouping on each re-render call.

valerybugakov replied:
I see. If it's not critical for this code to be optimized, consider removing memorization.

valerybugakov reviewed (approved):
UI looks great! Thanks for putting extra effort into keeping it more consistent.

vovakulikov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/DashboardSelect.module.scss:
```diff
@@ -0,0 +1,45 @@
+
+    &__group-label {
+        font-weight: normal;
+        font-size: 0.75rem;
```
Yeah. By default, reach ui adds `font-weight: bold` to the group label element. About font size, I changed font size according to sized in design layouts.

vovakulikov commented on client/web/src/insights/pages/dashboards/components/dashboard-select/DashboardSelect.module.scss:
```diff
@@ -0,0 +1,45 @@
+@import '~@reach/listbox/styles.css';
+
+.listbox {
+    &__button {
```
Noted. I've changed css classes. Thanks

vovakulikov commented:
@AlicjaSuska I'm about to merge this PR. Let's fix the remaining problems (like badge color and selected option bg color) in the next PR about the dashboard page https://github.com/sourcegraph/sourcegraph/pull/22474

For now, I applied bg-color-2 for the selected option in order that the selected option color won't overlap with badge secondary color.

AlicjaSuska commented:
Thank you @vovakulikov!
1. It's actually the other way around regarding backgrounds :)The badge should use bg-2 and the selected item background should be bg-3 (please, see the [design](https://www.figma.com/file/g5lSlPBnHt00wCogjRFnt5/RFC-398-Easily-Viewing-and-Sharing-Code-Insights-WIP?node-id=846%3A68)). Could you change that, please? Thank you!

2. Thanks, that's great
3. Awesome!
4. I've discussed the badge placement with @quinnkeast . @vovakulikov Could you please move them close to the name, as they are in the design? We came to the conclusion that by moving them to the right, they are over-emphasized visually. Thank you for exploring and thank you in advance for making this change!

vovakulikov commented:
Thanks @AlicjaSuska 

I've changed the color for the badge and selected option components. One concern here is that the bg-2 color isn't a standard color for a badge. For instance, most close to its color is badge-secondary variation and it uses bg-3. 

At the moment I applied colors according to figma layouts and your review but we need to reconsider the badge component and colors for it at some point.
